File: f_sorted_naive.py
# run: python f_sorted_naive.py < input/f5.in

# Command line to generate random input
# {echo 100000; paste -d ' ' <(gshuf -r -i 0-1000 -n 100000) <(gshuf -r -i 0-1000 -n 100000) } > input/f5.in

# Interestingly, this is the only python solution that I managed
# to squeeze into 3s time limit. Actually the longest test case took 2.995s
# so I was very lucky :)
# It uses sorted list to store and count points left to the current.
# Though search in this list takes O(log N), deletion items from it
# still takes O(N), so worst case is O(N*N).
# I tried to implement binary tree as in f_fastest.cpp, but surprisingly
# python version of it was even slower than this version and got TL

from bisect import bisect_left, bisect_right

n = int(input())
t, y = zip(*sorted(tuple(map(float, input().split())) for i in range(n)))
y_sorted = sorted(y)

nom = denom = 0
i = n - 1

while i >= 0:
    j = i
    while j >= 0 and t[i] == t[j]:
        l = bisect_left(y_sorted, y[j])
        y_sorted.pop(l)
        j -= 1

    for k in range(j + 1, i + 1):
        l = bisect_left(y_sorted, y[k])
        r = bisect_right(y_sorted, y[k])
        nom += l + (r - l) / 2
        denom += j + 1

    i = j

print(nom / denom)

File: f_sorted_naive.cpp
/*
  This is a bit better than naive solution. Here we use sorted vector
  to store points left to the current. It allows us to quickly in O(log N)
  count number of points with Y-coord less than current, but nevertheless
  insertion in this vector can take O(N) in worst case. So worst case
  is still O(N*N) as in naive solution, but even so it passes all tests
  and the longest one takes 1.6s
*/

#include <iostream>
#include <vector>
#include <algorithm>

using namespace std;

int main () {
    ios_base::sync_with_stdio(false);
    cin.tie(0);
    int n;
    cin >> n;
    vector<pair<double, double>> a(n);
    vector<double> y;
    for (int i = 0; i < n; i ++) {
        cin >> a[i].first;
        cin >> a[i].second;
    }

    sort(a.begin(), a.end(), [](pair<double, double> &x, pair<double, double> &y) {
      return x.first < y.first;
    });

    double nom = 0, denom = 0;

    for (int i = 0; i < n;) {
        int j = i;
        while (j < n && a[i].first == a[j].first) {
            auto l = lower_bound(y.begin(), y.end(), a[j].second) - y.begin();
            auto r = upper_bound(y.begin(), y.end(), a[j].second) - y.begin();
            nom += l + (r - l) / 2.0;
            denom += i;
            j ++;
        }

        for (int k = i; k < j; k ++) {
            auto l = lower_bound(y.begin(), y.end(), a[k].second);
            y.insert(l, a[k].second);
        }

        i = j;
    }

    cout << nom / denom;
}